src/libsyntax/parse/parser.rs

```diff
@@ -797,7 +797,18 @@ impl<'a> Parser<'a> {
         let mut err = self.struct_span_err(self.span,
                                            &format!("expected identifier, found {}",
                                                     self.this_token_descr()));
-        if let Some(token_descr) = self.token_descr() {
+        if let (true, token::Ident(ref s, false), true) = (
+            self.span.rust_2018(),
+            &self.token,
+            self.token.is_used_keyword() || self.token.is_unused_keyword(),
```

Comment:
The right condition here is `ident.is_reserved() && !ident.is_path_segment_keyword() && ident.name != keywords::Underscore.name()` because path segment keywords like `self` or `super` cannot be raw.